lib/msprime.c

```diff
@@ -807,53 +807,147 @@ msp_verify_segments(msp_t *self, bool verify_breakpoints)
+
+    free(self);
+}
+
+/* Find the number of segments that overlap at the given position */
+static uint32_t
+overlap_counter_overlaps_at(overlap_counter_t *self, double pos)
+{
+    assert(pos >= 0 && pos < self->seq_length);
+    segment_t *curr_overlap = self->overlaps;
+    while (curr_overlap->next != NULL) {
+        if (curr_overlap->left <= pos && pos < curr_overlap->right) {
+            break;
+        }
+        curr_overlap = curr_overlap->next;
+    }
+
+    return (uint32_t) curr_overlap->value;
+}
+
+/* Split the segment at breakpoint and add in another segment
+ * from breakpoint to seg.right. Set the original segment's
+ * right endpoint to breakpoint.
+ */
+static void
+overlap_counter_split_segment(segment_t *seg, double breakpoint)
+{
+    segment_t *right_seg = malloc(sizeof(segment_t));
+    right_seg->prev = NULL;
+    right_seg->next = NULL;
+    right_seg->left = breakpoint;
+    right_seg->right = seg->right;
+    right_seg->value = seg->value;
+    right_seg->population_id = 0;
+    right_seg->label = 0;
+
+    if (seg->next != NULL) {
+        right_seg->next = seg->next;
+        seg->next->prev = right_seg;
+    }
+    right_seg->prev = seg;
+    seg->next = right_seg;
+    seg->right = breakpoint;
+}
+
+/* Increment the number of segments that span
+ * [left, right], creating additional intervals if necessary.
+ */
+static void
+overlap_counter_increment_interval(overlap_counter_t *self, double left, double right)
+{
+    segment_t *curr_interval = self->overlaps;
+    while (left < right) {
+        if (curr_interval->left == left) {
+            if (curr_interval->right <= right) {
+                curr_interval->value++;
+                left = curr_interval->right;
+                curr_interval = curr_interval->next;
+            } else {
+                overlap_counter_split_segment(curr_interval, right);
+                curr_interval->value++;
+                break;
+            }
+        } else {
+            if (curr_interval->right < left) {
+                curr_interval = curr_interval->next;
+            } else {
+                overlap_counter_split_segment(curr_interval, left);
+                curr_interval = curr_interval->next;
+            }
+        }
+    }
+}
+
 static void
 msp_verify_overlaps(msp_t *self)
 {
     avl_node_t *node;
     node_mapping_t *nm;
     segment_t *u;
-    uint32_t j, k, count, label;
-    double left, right;
-    size_t remaining_samples = self->num_sampling_events - self->next_sampling_event;
-    /* We check for every locus, so obviously this rules out large numbers
-     * of loci. This code should never be called except during testing,
-     * so we don't need to recover from malloc failure.
-     */
-    uint32_t *overlaps = calloc((uint32_t) self->sequence_length, sizeof(uint32_t));
+    uint32_t j, label;
+    int remaining_samples = (int) (self->num_sampling_events - self->next_sampling_event);
+    overlap_counter_t *counter = overlap_counter_alloc(
```

Comment:
If we make this 
```
overlap_counter_t counter;

overlap_counter_alloc(&counter, self->sequence_length, remaining_samples);
```
then we keep the conventional API structure and avoid a malloc.